bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/TomlProvider.java

```diff
@@ -104,12 +107,76 @@
 
     TomlProvider(Module rootModule, Set<Module> moduleSet) {
         this.rootModule = rootModule;
-        this.moduleSet = moduleSet;
+        this.moduleInfo = new ModuleInfo(moduleSet);
     }
 
     @Override
     public void initialize() {
-        analyseModuleAmbiguity();
+        moduleInfo.analyseModules(rootModule);
+    }
+
+    @Override
+    public List<ConfigException> complete() {
+        if (tomlNode == null) {
+            return new ArrayList<>();
+        }
+        List<ConfigException> exceptionList = validateUnusedNodes(tomlNode, "");
+        visitedNodes.clear();
+        return exceptionList;
+    }
+
+    private List<ConfigException> validateUnusedNodes(TomlTableNode baseNode, String moduleHeader) {
+        List<ConfigException> exceptionList = new ArrayList<>();
+        for (Map.Entry<String, TopLevelNode> nodeEntry : baseNode.entries().entrySet()) {
+            TomlNode node = nodeEntry.getValue();
+            String lineRange = getLineRange(node);
+            String entryKey = nodeEntry.getKey();
+            boolean isInvalidNode = invalidTomlLines.contains(node.location().lineRange());
+            switch (node.kind()) {
+                case KEY_VALUE:
+                    if (!visitedNodes.contains(node) && !isInvalidNode) {
+                        exceptionList
+                                .add(new ConfigException(CONFIG_TOML_UNUSED_VALUE, lineRange, moduleHeader + entryKey));
```

Comment:
Creating a list of exceptions and added to the list, in this case, is not nice. Shall we change the complete method to pass the diagnostics log?
 void complete(RuntimeDiagnosticLog diagnosticLog);

Reply:
+1